udev: do not link with libudev

yuwata opened the pull request.

bluca commented:
CentOS failure is https://github.com/systemd/systemd/issues/17963 and bionic-* times out in NUMA-POLICY

bluca commented:
So the goal is to avoid linking a bunch of binaries with libudev, but instead to have the minimum required functions in libsystemd-shared, to reduce overall binary size, right?
The diff looks good to me, and no objections if this is the goal - let's wait a little and see if others want to chime in

poettering commented on src/udev/udev-builtin.h:
```diff
@@ -37,6 +37,10 @@ typedef struct UdevBuiltin {
 
+#define UDEV_NAME_SIZE   512
+#define UDEV_PATH_SIZE  1024
+#define UDEV_LINE_SIZE 16384
```
isn't that something to move src/shared/udev-util.h, too? It doesn't really have anything to do with builtins, no?

poettering commented on src/shared/udev-util.c:
```diff
@@ -536,3 +538,20 @@ int udev_resolve_subsys_kernel(const char *string, char *result, size_t maxsize,
 }
+
+int udev_queue_is_empty(void) {
+        return access("/run/udev/queue", F_OK) < 0;
```
maybe take the opportunity here to fix error propagation on this?
i.e.

```c
int udev_queue_is_empty(void) {
        return access("/run/udev/queue") < 0 ? 
                (errno == ENOENT ? true : -errno) : false;
```

And then if the other errors really should be treated equal to "is empty" handle that in the caller?

poettering reviewed (commented):
Love the clean-ups. Some minor suggestions.

yuwata commented:
@bluca and @poettering Thank you for your review. I've addressed the above suggestions. Setting the green label.

> So the goal is to avoid linking a bunch of binaries with libudev, but instead to have the minimum required functions in libsystemd-shared, to reduce overall binary size, right?

Right. Though, I am not sure about the binary size change. But I hope it is reduced slightly.

poettering commented:
> So the goal is to avoid linking a bunch of binaries with libudev, but instead to have the minimum required functions in libsystemd-shared, to reduce overall binary size, right?

I personally don't care too much about the binary size I must say. But I am very keen on keeping the codebase clean, and I think relegating "libudev" into "this is legacy mode", and thus not linking against it anymore from our own code is inherently a good thing.
